api/v3/rpc.go

```diff
@@ -83,32 +84,43 @@ func (s *AncestryServer) PostAncestry(ctx context.Context, req *pb.PostAncestryR
 	}
 
 	builder := clair.NewAncestryBuilder(clair.EnabledDetectors())
-	for _, layer := range req.Layers {
-		if layer == nil {
-			err := status.Error(codes.InvalidArgument, "ancestry layer is invalid")
-			return nil, err
-		}
-
-		if layer.GetHash() == "" {
-			return nil, status.Error(codes.InvalidArgument, "ancestry layer hash should not be empty")
-		}
-
-		if layer.GetPath() == "" {
-			return nil, status.Error(codes.InvalidArgument, "ancestry layer path should not be empty")
+	layerMap := map[string]*database.Layer{}
+	g, analyzerCtx := errgroup.WithContext(ctx)
+	for i := range req.Layers {
+		layer := req.Layers[i]
+		if _, ok := layerMap[layer.Hash]; !ok {
+			layerMap[layer.Hash] = nil
+			if layer == nil {
+				err := status.Error(codes.InvalidArgument, "ancestry layer is invalid")
+				return nil, err
+			}
+
+			if layer.GetHash() == "" {
+				return nil, status.Error(codes.InvalidArgument, "ancestry layer hash should not be empty")
+			}
+
+			if layer.GetPath() == "" {
+				return nil, status.Error(codes.InvalidArgument, "ancestry layer path should not be empty")
+			}
+
+			g.Go(func() error {
+				clairLayer, err := clair.AnalyzeLayer(analyzerCtx, s.Store, layer.Hash, req.Format, layer.Path, layer.Headers)
+				if err != nil {
+					return err
+				}
+
+				layerMap[layer.Hash] = clairLayer
```

Comment:
This needs a lock even though the keys are independent because of races when the runtime is expanding the size of the map.

Even if we added the size to the make,`layerMap := make(map[string]*database.Layer, len(req.Layers))`, I wouldn't trust this race without a lock.

Reply:
makes sense